### PR TITLE
Fix delta lake CDF entries for MERGE queries

### PR DESCRIPTION
**Pull request by homar**
## Description
Before this change MERGE that made both UPDATE and DELETE could create CDF entries with rows marked as delete when in fact there were update_preimage and other way around

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# Delta Lake
* Fix change data feed entries for `MERGE` queries. ({issue}`16127`)
```

### CODE REVIEW

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java [hunk 1]
```diff
@@ -399,6 +402,7 @@ private Optional<DataFileInfo> rewriteParquetFile(Path path, FileDeletion deleti
             throws IOException
     {
         LongBitmapDataProvider rowsToDelete = deletion.rowsToDelete;
+        LongBitmapDataProvider rowsDeletedBecauseOfUpdate = deletion.rowsDeletedBecauseOfUpdate();
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java:405]: Can we avoid marking update preimage rows in both Bitmaps?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java [hunk 2]
```diff
@@ -410,21 +414,34 @@ private Optional<DataFileInfo> rewriteParquetFile(Path path, FileDeletion deleti
...
                 int[] retained = new int[positionCount];
                 int[] deleted = new int[positionCount];
+                int[] deletedBecauseOfUpdate = new int[positionCount];
+                int[] deletedBecauseOfDelete = new int[positionCount];
                 int retainedCount = 0;
                 int deletedCount = 0;
+                int deletedBecauseOfUpdateCount = 0;
+                int deletedBecauseOfDeleteCount = 0;
                 for (int position = 0; position < positionCount; position++) {
                     if (!rowsToDelete.contains(filePosition)) {
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java:424]: I'd rephrase these as:

```
if (rowsToDelete.contains(filePosition)) {
    ...
}
else if (rowsDeletedBecauseOfUpdate.contains(filePosition)) {
    ...
}
else {
    // retain ...
}

**Review by alexjo2144 — COMMENTED** [review]
Good catch

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java (hunk 1 above)
Comment (homar) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java:405]: i don't think so, we need to carry information if particular position deletion came from DELETE or from UPDATE, instead of a bitmap a normal Map could be used that would store the reason for the deletion but I assumed it is more efficient to use 2 bitmaps.
I can change to one Map if you think it is better

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java (hunk 2 above)
Comment (homar) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java:424]: But that would be incorrect. I am interested in rows that will be deleted  I just need to split them into 2 types

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java (hunk 1 above)
Comment (alexjo2144) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java:405]: Two Bitmaps is good, but you have redundant information for updated rows, the positions for both `rowsToDelete` and `rowsDeletedBecauseOfUpdate` are set. I'd expect you to set one or the other, not both.

Comment (homar) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java:405]: ok I think I figured out your point, and now the second if would probably also work the way you wrote it, I will give it a shot. thanks

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java (hunk 2 above)
Comment (homar) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java]: Ok my bad, I figured it out and it works. Thanks

**Comment by findinpath** [comment]
Build is red (error-prone-checks)

```
 Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.9.0:compile (default-compile) on project trino-delta-lake: Compilation failure: Compilation failure: 
Error:  /home/runner/work/trino/trino/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java:[418,23] [UnusedVariable] The local variable 'deleted' is never read.
```

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java [hunk 3]
```diff
@@ -410,21 +416,30 @@ private Optional<DataFileInfo> rewriteParquetFile(Path path, FileDeletion deleti
                 int positionCount = page.getPositionCount();
                 int[] retained = new int[positionCount];
                 int[] deleted = new int[positionCount];
+                int[] deletedBecauseOfUpdate = new int[positionCount];
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java:419]: We created before this change one array with `positionCount` items. Now we create two arrays with `positionCount` items. Is this valid?

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksChangeDataFeedCompatibility.java [hunk 4]
```diff
@@ -379,6 +379,68 @@ public void testMergeDeleteIntoTableWithCdfEnabled()
...
+    public void testMergeMixedDeleteAndUpdateIntoTableWithCdfEnabled()
+    {
+        String tableName1 = "test_merge_mixed_delete_and_update_into_table_with_cdf_" + randomNameSuffix();
+        String tableName2 = "test_merge_mixed_delete_and_update_into_table_with_cdf_data_table_" + randomNameSuffix();
+        try {
+            onDelta().executeQuery("CREATE TABLE default." + tableName1 + " (nationkey INT, name STRING, regionkey INT) " +
+                    "USING DELTA " +
+                    "LOCATION 's3://" + bucketName + "/databricks-compatibility-test-" + tableName1 + "'" +
+                    "TBLPROPERTIES (delta.enableChangeDataFeed = true)");
+            onDelta().executeQuery("CREATE TABLE default." + tableName2 + " (nationkey INT, name STRING, regionkey INT) " +
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksChangeDataFeedCompatibility.java:392]: this table does not need to enable cdf

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksChangeDataFeedCompatibility.java [hunk 5]
```diff
@@ -379,6 +379,68 @@ public void testMergeDeleteIntoTableWithCdfEnabled()
         }
     }
 
+    @Test(groups = {DELTA_LAKE_DATABRICKS, DELTA_LAKE_EXCLUDE_73, PROFILE_SPECIFIC_TESTS})
+    public void testMergeMixedDeleteAndUpdateIntoTableWithCdfEnabled()
+    {
+        String tableName1 = "test_merge_mixed_delete_and_update_into_table_with_cdf_" + randomNameSuffix();
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksChangeDataFeedCompatibility.java:385]: `tableName1` -> `targetTableName`
`tableName2` -> `sourceTableName`

as in the syntax of `MERGE` https://trino.io/docs/current/sql/merge.html

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksChangeDataFeedCompatibility.java [hunk 6]
```diff
@@ -379,6 +379,68 @@ public void testMergeDeleteIntoTableWithCdfEnabled()
...
+            onDelta().executeQuery("INSERT INTO default." + tableName2 + " VALUES (2, 'nation2', 20000)");
+            onDelta().executeQuery("INSERT INTO default." + tableName2 + " VALUES (3000, 'nation3000', 3000)");
+            onDelta().executeQuery("INSERT INTO default." + tableName2 + " VALUES (4, 'nation4000', 4000)");
+
+            onTrino().executeQuery("MERGE INTO delta.default." + tableName1 + " cdf USING delta.default." + tableName2 + " n " +
+                    "ON (cdf.nationkey = n.nationkey) " +
+                    "WHEN MATCHED AND cdf.nationkey = 2 " +
+                    "THEN DELETE " +
+                    "WHEN MATCHED AND cdf.nationkey > 2 " +
+                    "THEN UPDATE SET nationkey = (cdf.nationkey + n.nationkey + n.regionkey) " +
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksChangeDataFeedCompatibility.java:412]: optional nit for getting the test easier to read

Let's rather update the field you have now named as `regionkey` to easier follow what gets changes by having the `nationkey` as key.

I'd go with a `page_views (pagekey, url, views)` sample which is easier to follow (when merging a page with the same id, the views get summed).

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksChangeDataFeedCompatibility.java [hunk 7]
```diff
@@ -379,6 +379,68 @@ public void testMergeDeleteIntoTableWithCdfEnabled()
...
+            onDelta().executeQuery("INSERT INTO default." + tableName1 + " VALUES (3, 'nation3', 300)");
+            onDelta().executeQuery("INSERT INTO default." + tableName1 + " VALUES (4, 'nation4', 400)");
+
+            onDelta().executeQuery("INSERT INTO default." + tableName2 + " VALUES (1000, 'nation1000', 1000)");
+            onDelta().executeQuery("INSERT INTO default." + tableName2 + " VALUES (2, 'nation2', 20000)");
+            onDelta().executeQuery("INSERT INTO default." + tableName2 + " VALUES (3000, 'nation3000', 3000)");
+            onDelta().executeQuery("INSERT INTO default." + tableName2 + " VALUES (4, 'nation4000', 4000)");
+
+            onTrino().executeQuery("MERGE INTO delta.default." + tableName1 + " cdf USING delta.default." + tableName2 + " n " +
+                    "ON (cdf.nationkey = n.nationkey) " +
```
[the remark this answers is not in the document]
Comment (homar) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksChangeDataFeedCompatibility.java:408]: hmm ok but I am not a big fun, the fact they don't support interpolation is a no go for me

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java (hunk 3 above)
Comment (homar) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java:419]: We created 2 arrays with positionCount:
```
int[] retained = new int[positionCount];
int[] deleted = new int[positionCount];
```
now we create three, this is valid, these arrays are only used to get specific positions from page

Comment (findinpath) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java:418]: Why not use `rowsDeletedBecauseOfUpdate.getLongCardinality()` instead?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java [hunk 8]
```diff
@@ -520,30 +533,29 @@ public void abort()
...
         {
-            return rowsToDelete;
+            return rowsDeletedBecauseOfDelete;
         }
 
-        public String getCdfOperation()
+        public LongBitmapDataProvider rowsDeletedBecauseOfUpdate()
         {
-            return cdfOperation;
+            return rowsDeletedBecauseOfUpdate;
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java:558]: nit: How about renaming to like below? 
* `rowsDeletedBecauseOfDelete` → `rowsDeletedByDelete`
* `rowsDeletedBecauseOfUpdate` → `rowsDeletedByUpdate`

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksChangeDataFeedCompatibility.java [hunk 9]
```diff
@@ -379,6 +379,71 @@ public void testMergeDeleteIntoTableWithCdfEnabled()
...
+                    MERGE INTO delta.default.%s cdf USING delta.default.%s n
+                    ON (cdf.page_id = n.page_id)
+                    WHEN MATCHED AND cdf.page_id = 2
+                        THEN DELETE
+                    WHEN MATCHED AND cdf.page_id > 2
+                        THEN UPDATE SET views = (cdf.views + n.views)
+                    WHEN NOT MATCHED
+                        THEN INSERT (page_id, page_url, views) VALUES (n.page_id, n.page_url, n.views)""",
+                            targetTableName,
+                            sourceTableName));
```
Comment (ebyhr) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksChangeDataFeedCompatibility.java:417]: nit: Only two variables are used here. How about using string concatenation for SQL syntax highlight in IntelliJ?

Comment (homar) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksChangeDataFeedCompatibility.java:417]: This is what I had but Marius insisted on Text Block. I also prefer concat

**Comment by homar** [comment]
@ebyhr could you please run this one with secrets ?

**Comment by ebyhr** [comment]
/test-with-secrets sha=f1c89aadc1ab6472d4c292a04e55daa2c8a511bb

**Review by findinpath — APPROVED** [review]

**Comment by ebyhr** [comment]
CI is running in https://github.com/trinodb/trino/actions/runs/4201339088/jobs/7288250233
